### ISSUE TITLE
Hitman 3 Ray Tracing

### ISSUE THREAD

**Issue by JonLit** [HansKristian-Work/vkd3d-proton#1148]
Hi,
Ray Tracing isn't working for me in H3 (no visual difference when any of the ray tracing options are enabled or not)
H3 version: 3.110.1
I'm using GE-Proton7-21, also doesn't work with other standard Proton versions and GE-Proton7-20,
with these options: `VKD3D_CONFIG=dxr11 RADV_PERFTEST=rt %command%`
I have all other game settings on maximum, but I don't think these should have an impact on RT working or not
System:
rx 6700xt
i5-12600k
Garuda Linux (Arch-based)
mesa 22.1.2-1

If this does not belong here, please tell me, I'm very sorry in that case

**Comment by jhu54** [comment]
With Hitman 3, I am guessing you enabled Ray Tracing in the launcher options, clicked save then launched the game with those settings?

Did you check the graphics settings in the ingame options to see if you can enable ray tracing there? With 6700xt/11600K, git pull of VKD3D from last night and Mesa 22.2 Devel, I can enable ray tracing in the launch options area but ingame, the option is disabled. It appears that not everything is ready for ray tracing in Hitman 3 under linux just yet...

**Comment by benbeshara** [comment]
I'm having issues with RT in Hitman 3 as well. I can enable it and DLSS in the launcher (`VKD3D_CONFIG=dxr DXVK_FEATURE_LEVEL=12_1` etc. in env variables) and the frame rate absolutely tanks so, something is happening - but RT reflections aren't displaying (I honestly can't tell if lighting/shadows/etc are working, the frame rate stabalises when reflections are turned off and I'm not sure what to make of that?). Cyberpunk 2077 RT is working with the same set-up. Like jhu54 above the settings are disabled in-game, I'm not sure if this is normal as I don't run Windows anymore to test.

- RTX 3060Ti/Ryzen 5 3-series
- Arch Linux
- nVidia 515 drivers
- Latest Wine-GE via Heroic Launcher
- [rt-subobject-association](https://github.com/HansKristian-Work/vkd3d-proton/tree/rt-subobject-association) branch of vkd3d (this branch made a huge difference in being able to enable and use RT in Cyberpunk)
- DXVK NVAPI DLL

I'm going to download the demo of the Steam version and see if Proton-GE gives me a different result

**Comment by JonLit** [comment]
> With Hitman 3, I am guessing you enabled Ray Tracing in the launcher options, clicked save then launched the game with those settings?
> 
> Did you check the graphics settings in the ingame options to see if you can enable ray tracing there? With 6700xt/11600K, git pull of VKD3D from last night and Mesa 22.2 Devel, I can enable ray tracing in the launch options area but ingame, the option is disabled. It appears that not everything is ready for ray tracing in Hitman 3 under linux just yet...

You're right, I enabled it in the launcher settings, ingame I can't change it (the RT options are greyed out) but they say they're enabled...

**Comment by Etaash-mathamsetty** [comment]
there were a bunch of rt related changes that got merged 2 days ago, maybe try this again?

**Comment by benbeshara** [comment]
No dice here :( Same result as before

**Comment by deescuderoo** [comment]
In my case (RTX 3080 Ti) the option in the launcher is greyed out. The game does detect the right GPU, and I'm using this as launcher options:

> VKD3D_CONFIG=dxr11 DXVK_FEATURE_LEVEL=12_1 PROTON_ENABLE_NVAPI=1 PROTON_HIDE_NVIDIA_GPU=0 %command%

Any ideas?

**Comment by davigamer987** [comment]
Tried with VKD3D_CONFIG=dxr11 RADV_PERFTEST=rt DXVK_FEATURE_LEVEL=12_1, In the launcher i could enable ray tracing but in game it showed that only ray tracing was enabled (Ray traced Sun Shadows and reflections were off, and the options were greyed out)

RX 6950xt, R7 5800X

**Comment by looncraz** [comment]
Same results here as JonLit, RT ON, but can't change settings in game, and no visual change or FPS drop

Relevant Hardware:

Ryzen 9 7950X
32GB DDR5-5600 CL32
Radeon 6700XT, stock

Software:

Xubuntu 22.04, fully updated
Mesa 23.0.0-devel (from today)
kernel 6.0.0, latest firmware
compiz 0.9.14.1 (yeah, I'm insane, XFCE+compiz, LOL)

Steam, GE-Proton7-42 (tried all others I could)

`VKD3D_FEATURE_LEVEL=12_2 RADV_PERFTEST=rt VKD3D_CONFIG=dxr11 MANGO_HUD=1 %command% -dx12`

**Comment by lilxandrias** [comment]
Same here. 

Ryzen 5 1600
RX 6650 XT
Arch Linux
Wine-GE-Proton7-35

**Comment by gianmarcotoso** [comment]
Same issue, launching with `VKD3D_CONFIG=dxr11 DXVK_FEATURE_LEVEL=12_1 PROTON_ENABLE_NVAPI=1 PROTON_HIDE_NVIDIA_GPU=0 gamemoderun mangohud %command%` and RT is disabled in the game's options menu. I cannot access the options through the launcher, clicking "Options" just launches the game instead.

I've tried forcing RT on by editing the registry with protontricks but I don't see RT working in-game.
 
13900k
RTX 3090
Arch Linux
GE-Proton-7-43 (Also tried Proton 7.0-5 and Experimental)

**Comment by lilxandrias** [comment]
I am experiencing same issues with Resident Evil RE engine based games too.

**Comment by Blisto91** [comment]
Toyed a bit around with this (Hitman 3 RT) and it doesn't seem to be a vkd3d-proton issue. I can enable RT fine ingame on Windows with vkd3d-proton when i also give it a bogus vendorId e.g. `dxgi.customVendorId = ffff` which you would put into a `dxvk.conf` file next to the exe ( HITMAN 3/Retail/ ). Tested with a 7900XTX.
This also works on Linux if i do the same thing. 

I am unsure what info it is looking for without this workaround that doesn't fulfills it's requirements for AMD since setting `amd_ags_x64` to native doesn't help. Windows Driver version reported through there is relatively up to date.

**Comment by davigamer987** [comment]
> Toyed a bit around with this (Hitman 3 RT) and it doesn't seem to be a vkd3d-proton issue. I can enable RT fine ingame on Windows with vkd3d-proton when i also give it a bogus vendorId e.g. `dxgi.customVendorId = ffff` which you would put into a `dxvk.conf` file next to the exe ( HITMAN 3/Retail/ ). Tested with a 7900XTX. This also works on Linux if i do the same thing.
> 
> I am unsure what info it is looking for without this workaround that doesn't fulfills it's requirements for AMD since setting `amd_ags_x64` to native doesn't help. Windows Driver version reported through there is relatively up to date.

If I just set that in the dxvk conf it works? Do I need anything else in that file? Any command line ARGS needed?

**Comment by Blisto91** [comment]
Well works and works. My GPU hung and second time game black screened on loading

I dunno what GPU you have. If you have amd you'd need `VKD3D_CONFIG=dxr11 RADV_PERFTEST=rt %command%`
For Nvidia just `VKD3D_CONFIG=dxr11 %command%`
Unsure if Nvapi needs to be enabled on Nvidia for RT even if we report another GPU vendor.

**Comment by frozen-sea** [comment]
Just ran across this issue when trying to get RT working in the demo version. Simply having RT features enabled for vkd3d leaves the RT option in-game greyed out, but setting a bogus vendor id as suggested by @Blisto91 changes that. With the workaround in place both RT reflections and shadows can be toggled in game and render correctly. Of course, that means losing access to DLSS.

4070 Ti, driver 535.86.05
Nobara 38, kernel 6.3.12
GE-Proton8-9

**Comment by davigamer987** [comment]
RT is still broken with the dummy vendor ID, now when i open the game i get a vkCreateRayTracingPipelinesKHR error

**Comment by pixelcluster** [comment]
> when i open the game i get a vkCreateRayTracingPipelinesKHR error
(RX 6950XT)

This is probably a RADV issue. In current stable Mesa versions (up to 23.1.x) that is sort of to be expected. These issues should be gone in 23.2, but that isn't released yet as of now. Best thing is probably to wait until 23.2 is out (shouldn't be too long until it is released) and then try again.

**Comment by davigamer987** [comment]
> > when i open the game i get a vkCreateRayTracingPipelinesKHR error
> > (RX 6950XT)
> 
> This is probably a RADV issue. In current stable Mesa versions (up to 23.1.x) that is sort of to be expected. These issues should be gone in 23.2, but that isn't released yet as of now. Best thing is probably to wait until 23.2 is out (shouldn't be too long until it is released) and then try again.

same issue with 23.3.0

**Comment by Blisto91** [comment]
Post a log

**Comment by ADHDefy** [comment]
I'm having a related issue where the Ray Tracing option is just greyed out for me. I'm on Fedora 38, I have an RTX 2070s, and ray tracing works in other titles just fine. I've tried the latest Proton, Experimental, and GE 8-13, along with every combo of these launch options: `PROTON_HIDE_NVIDIA_GPU=0 PROTON_ENABLE_NVAPI=1 VKD3D_CONFIG=dxr DXVK_FEATURE_LEVEL=12_1 gamemoderun %command%'

It just won't let me enable it.

**Comment by SeongGino** [comment]
> I'm having a related issue where the Ray Tracing option is just greyed out for me. I'm on Fedora 38, I have an RTX 2070s, and ray tracing works in other titles just fine. I've tried the latest Proton, Experimental, and GE 8-13, along with every combo of these launch options: `PROTON_HIDE_NVIDIA_GPU=0 PROTON_ENABLE_NVAPI=1 VKD3D_CONFIG=dxr DXVK_FEATURE_LEVEL=12_1 gamemoderun %command%`
> 
> I have enabled DLSS and that was no problem, so one of the Nvidia features is working.
> 
> It just won't let me enable it in the launcher or in the game menu.

That is literally this whole issue in a nutshell. Welcome to the club.

**Comment by xanscale** [comment]
Ryzen 5600X + RX7600
Linux 6.5.6
Mesa 23.2.1

.../HITMAN 3/Retail/dxvk.conf 
`dxgi.customVendorId = ffff`

VKD3D_CONFIG=dxr11 gamemoderun %command% -skip_launcher

Ray tracing works fine (no artifact) but incredible drop performance (75 to 5 fps) but i suppose is normal with my gpu

**Comment by davigamer987** [comment]
> Ryzen 5600X + RX7600 Linux 6.5.6 Mesa 23.2.1
> 
> .../HITMAN 3/Retail/dxvk.conf `dxgi.customVendorId = ffff`
> 
> VKD3D_CONFIG=dxr11 gamemoderun %command% -skip_launcher
> 
> Ray tracing works fine (no artifact) but incredible drop performance (75 to 5 fps) but i suppose is normal with my gpu

Interesting, maybe my issue is 6xxx only, I can't even open the game with RT enabled

**Comment by xanscale** [comment]
@davigamer987 mesa version ? (23.2 is the first that enable RT for RADV)

**Comment by davigamer987** [comment]
> @davigamer987 mesa version ? (23.2 is the first that enable RT for RADV)

23.3.0

**Comment by Blisto91** [comment]
Can you post a proton or wine log please

**Comment by Blisto91** [comment]
Can AMD users verify that the Ray Tracing option is no longer grayed out without a override on Proton Experimental Bleeding Edge? (mesa 23.2.1+)

If Nvidia users could give it a recheck that would also be great.

**Comment by frozen-sea** [comment]
> If Nvidia users could give it a recheck that would also be great.

Still greyed out, even when passing VKD3D_CONFIG=dxr

On the upside, DLSS now works instead of turning the whole screen bright yellow

**Comment by spiffeeroo** [comment]
> Can AMD users verify that the Ray Tracing option ingame is no longer grayed out without a override on Proton Experimental Bleeding Edge? (mesa 23.2.1+). No vkd3d-proton configs should be needed either.

Ray tracing is no longer grayed out for me. Ray tracing, ray traced reflections, and sun shadows can be toggled. No commands in the Steam launcher. I have AMD 7900 XTX and 7800 XT cards and Mesa 23.2.1.

**Comment by SeongGino** [comment]
> Can AMD users verify that the Ray Tracing option ingame is no longer grayed out without a override on Proton Experimental Bleeding Edge? (mesa 23.2.1+). No vkd3d-proton configs should be needed either.
> 
> If Nvidia users could give it a recheck that would also be great.

If we're talking VKD3D 2.11, DLSS works but RT is still grayed out. 3060ti.

**Comment by Milor123** [comment]
I have the same problem and i have the RTX 4070  :/, anyone solution?

**Comment by Blisto91** [comment]
No the cause is not known yet.

**Comment by aguinet** [comment]
With nvidia GPUs, `DXVK_CONFIG="dxgi.customVendorId = ffff"` indeed enables RT and makes RT works in game. DLSS though, even if configurable, does not seem to change anything, as I'm getting the same performances/image quality in `Quality` or `Performance` mode.

**Comment by Blisto91** [comment]
This should now also work on Nvidia on Proton Bleeding Edge without any workarounds-

**Comment by deescuderoo** [comment]
Thanks, although for me the Ray Tracing option is still greyed out on Proton Experimental. My GPU is a RTX 3080 Ti. I tried both with 
```
VKD3D_CONFIG=dxr11 DXVK_FEATURE_LEVEL=12_1 PROTON_ENABLE_NVAPI=1 PROTON_HIDE_NVIDIA_GPU=0 %command%
```
and also without it, in both cases it is greyed out.

**Comment by Blisto91** [comment]
It isn't in Experimental yet. Only the Experimental Bleeding Edge beta branch

**Comment by aguinet** [comment]
I confirm that this works by building https://github.com/GloriousEggroll/proton-ge-custom by hand with https://github.com/jp7677/dxvk-nvapi/commit/42ec75b470ba6aa9b8c078f5528f59832920f7c4 for dxvk-nvapi. Indeed, unfortunately, the latest Proton-GE has been built one day before that patch landed in dxvk-nvapi...

You also need NVIDIA driver >= 550.54.14 to prevent XID CTX SWITCH TIMEOUT errors.

Thanks @jp7677 for the fix!

**Comment by Blisto91** [comment]
Is included with the new Proton 9 Experimental

**Comment by deescuderoo** [comment]
> Thanks, although for me the Ray Tracing option is still greyed out on Proton Experimental. My GPU is a RTX 3080 Ti. I tried both with
> 
> ```
> VKD3D_CONFIG=dxr11 DXVK_FEATURE_LEVEL=12_1 PROTON_ENABLE_NVAPI=1 PROTON_HIDE_NVIDIA_GPU=0 %command%
> ```
> 
> and also without it, in both cases it is greyed out.

Yep, I can confirm it works now with experimental. It's great!

**Comment by TheOverpassArsonist** [comment]
I can see it with proton experimental, but I'm not sure it's working properly. I have a 3080 and can otherwise put every setting at max and stay at a completely locked 120hz, but the second I turn on any ray tracing my performance dips to around the 40s. 

Obviously ray tracing is a fairly costly thing to enable, but given that the 30 series was released just a few months before Hitman 3 was and I'm playing at 1080p I don't think Ray Tracing should be over halving my frame rate like that. The only time I've ever had an FPS drop quite this drastic was with Ghostrunner and it was an issue with the game itself that was fixed by installing some of the updates. (IIRC I got it on GOG or something and whateverl auncher I was using didn't auto-install the updates)

I'm by no means a graphics pipeline expert, but locked 120fps with everything on absolute max taken down to 40fps with medium-high settings seems like it's more likely to be an issue than not. Can someone else check and see what performance they get with it enabled because a drop that drastic doesn't feel right to me.
